Issue/14128 fetch related posts on reader post details screen from vm + usecase

ashiagr opened the pull request:
Task #14128

This PR extracts only the "related posts fetch logic" from the reader post details fragment to useCase and viewModel making this part of code testable.

PS:
1. It currently targets `issue/14128-reader-related-posts` which will be merged to develop after implementing remaining sub-tasks from the linked task.
2. I'll try to refactor "related posts display related logic" in the next sub-task where I'll update the "related posts" design.

To test:

Test 1
- Open a wp.com Reader post.
- Notice that related posts are displayed as before at the bottom of the Reader post details screen.

Test 2
- Open a non wp.com Reader post.
- Notice that related posts are not displayed at the bottom of the Reader post details screen.

Test 3
- Open a wp.com Reader post with scrollable content.
- Enable Aeroplane mode.
- Scroll to bottom.
- Notice that related posts fetch request fails silently (matches behaviour prior to refactoring) and related posts are not shown.

Tests included in `ReaderPostDetailViewModelTest`, `ReaderFetchRelatedPostsUseCaseTest`.

PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ashiagr committed:
Add field to check if related posts fetch request succeeded

ashiagr committed:
Return event bus response on related posts request failure

ashiagr committed:
Refactor: Update only post actions to preserve related posts ui state on post action handle

ashiagr committed:
Refactor: Remove R.string import

ashiagr committed:
Refactor: Extract post.isWp condition to VM

ashiagr committed:
Add useCase to fetch reader related posts

ashiagr committed:
Refactor: Update only follow button ui state to preserve related posts ui state when post is followed

ashiagr committed:
Tests: Update ReaderPostDetailViewModel params

ashiagr committed:
Tests: Reword existing tests statements to given/when/then format and remove Arrange/Act/Assert comments

ashiagr committed:
Tests: Removed explicit type args, Assertions import

ashiagr committed:
Tests: Add tests for fetch related posts use case

ashiagr committed:
Request "related posts" from the view model

ashiagr committed:
Tests: Add tests for fetch related posts use case

ashiagr committed:
Merge branch 'issue/14128-reader-related-posts' into issue/14128-fetch-related-posts-from-vm-usecase

ashiagr committed:
Merge branch 'issue/14128-reader-related-posts' into issue/14128-fetch-related-posts-from-vm-usecase

ashiagr committed:
Fix: Baseline detekt for legacy code

ashiagr committed:
Fix: Do nothing on related posts result != success to match current behaviour

ashiagr committed:
Add bottom margin

ParaskP7 commented on WordPress/src/main/java/org/wordpress/android/ui/reader/usecases/ReaderFetchRelatedPostsUseCase.kt:
```diff
@@ -0,0 +1,78 @@
+            }
+            !networkUtilsWrapper.isNetworkAvailable() -> {
+                NoNetwork
+            }
```
Minor (🔍): Consider removing braces for the `AlreadyRunning` and `NoNetwork` when entries. This will save us a few lines and will make it a bit more readable, wdyt?

ashiagr replied:
In 16fb72f

ParaskP7 commented on WordPress/src/main/java/org/wordpress/android/ui/reader/viewmodels/ReaderPostDetailViewModel.kt:
```diff
@@ -179,6 +193,22 @@ class ReaderPostDetailViewModel @Inject constructor(
+            when (val fetchRelatedPostsState = readerFetchRelatedPostsUseCase.fetchRelatedPosts(sourcePost)) {
+                is FetchRelatedPostsState.AlreadyRunning,
+                is FetchRelatedPostsState.Failed.NoNetwork,
+                is FetchRelatedPostsState.Failed.RequestFailed -> { // Do Nothing
```
Minor (🔍): Instead of using braces, I usually do this trick to get to a one-liner result. I am sharing here in case you might like it as well:
```
is FetchRelatedPostsState.Failed.RequestFailed -> Unit // Do Nothing
```

PS: I actually created a helper function on my previous roles, like a `doNothing`, which is an alias to `Unit` and this way it can end up even easier to read and apprehend, without the need for commenting:

```
is FetchRelatedPostsState.Failed.RequestFailed -> doNothing
```

ashiagr replied:
Nice suggestion, in c4bcf98.

ParaskP7 commented on WordPress/src/main/java/org/wordpress/android/ui/reader/viewmodels/ReaderPostDetailViewModel.kt:
```diff
@@ -199,16 +229,54 @@ class ReaderPostDetailViewModel @Inject constructor(
         )
     }
 
+    private fun updateFollowButtonUiState(
```
Question (❓): Since I didn't dig too deeply into these changes, would you mind explaining it in a sentence? From the commit message I couldn't understand the full picture in terms of how it used be behave:

```
Refactor: Update only follow button ui state to preserve related posts ui state when post is followed
```

ashiagr replied:
I made this change because related posts are fetched separately and are not part of post model. Earlier when a post was followed, `ReaderPostDetailsUiState` was recreated using the modified post model. Since modified post model doesn't contain related posts, I chose to update just the follow button state and not recreating entire ui state using this modified post model otherwise related posts will be set to null in the ui state.

ParaskP7 replied:
Thanks so much Ashita for explaining all that to me, much much appreciated! 🙏

ParaskP7 commented on WordPress/src/main/java/org/wordpress/android/ui/reader/viewmodels/ReaderPostDetailViewModel.kt:
```diff
@@ -199,16 +229,54 @@ class ReaderPostDetailViewModel @Inject constructor(
+        _uiState.value = currentUiState.copy(headerUiState = updatedHeaderUiState)
+    }
+
+    private fun updatePostActions(post: ReaderPost) {
```
Question (❓): Since I didn't dig too deeply into these changes, would you mind explaining it in a sentence? From the commit message I couldn't understand the full picture in terms of how it used be behave:

```
Refactor: Update only post actions to preserve related posts ui state on post action handle
```

ashiagr replied:
Same as above

I made this change because related posts are fetched separately and are not part of post model. Earlier when a post action (like/ bookmark etc) were performed, `ReaderPostDetailsUiState` was recreated using the modified post model. Since modified post model doesn't contain related posts, I chose to update just the actions state and not recreating entire ui state using this modified post model.

ParaskP7 replied:
Thanks so much Ashita for explaining all that to me, much much appreciated! 🙏

ParaskP7 commented on WordPress/src/main/java/org/wordpress/android/ui/reader/viewmodels/ReaderPostDetailViewModel.kt:
```diff
@@ -199,16 +229,54 @@ class ReaderPostDetailViewModel @Inject constructor(
-        val actions: ReaderPostActions
-    )
+        val actions: ReaderPostActions,
+        /* related posts from the same site as the current post */
```
Suggestion (💡): Maybe these comments could be located closer to the source? Like `ReaderPostActions.handleRelatedPostsResponse(...)`?

ashiagr replied:
Comment already exists [in the source](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/ui/reader/actions/ReaderPostActions.java#L430).
`// split into posts from the passed site (local) and from across wp.com (global)`

It looks sufficient to me. Wdyt?

ParaskP7 replied:
Yeap, definitely, to me too. Why I comment on that is because maybe we should consider removing those from here. Wdyt? Do we really need them there?

ashiagr replied:
Got it, comment removed in d54dfd2.

ParaskP7 replied:
Many thanks once more @ashiagr !

Feel free to merge this PR whenever you are ready. 🌟

ParaskP7 commented on WordPress/src/test/java/org/wordpress/android/ui/reader/usecases/ReaderFetchRelatedPostsUseCaseTest.kt:
```diff
@@ -0,0 +1,77 @@
+
+        whenever(networkUtilsWrapper.isNetworkAvailable()).thenReturn(true)
+    }
+
```
Minor (🔍): Is it possible for us to also test the `AlreadyRunning` logic somehow?

ashiagr replied:
I tried to include a test but couldn't do it with event bus.  

We can test it manually by commenting below line [here](https://github.com/wordpress-mobile/WordPress-Android/pull/14142/files/087f3d6cee3ba37b0ff4ec1755a1ee5a59aee046#diff-7b33e01a0849528abbcbf8188cef615b72a21a6e2cf3656101b7bc8e0a454496R58):
` continuations[request] = null`

and fetching related posts twice from `ReaderPostDetailViewModel.onRelatedPostsRequested`.

ParaskP7 replied:
Cool, great, yes, having fun with event bus, what can you do... 😞 🤷‍♂️

ParaskP7 commented on WordPress/src/test/java/org/wordpress/android/ui/reader/viewmodels/ReaderPostDetailViewModelTest.kt:
```diff
@@ -298,12 +290,102 @@ class ReaderPostDetailViewModelTest {
+
+                viewModel.onRelatedPostsRequested(nonWpComPost)
+
+                verify(readerFetchRelatedPostsUseCase, times(0)).fetchRelatedPosts(readerPost)
```
Minor (🔍): I usually prefer the `never()` instead of `times(0)`. I am just mentioning that in case you weren't aware of it or you have a good reason I am not aware of using `times(0)` instead. Just sharing, nothing more, don't expect you to change it.

ParaskP7 reviewed (approved):
👋 @ashiagr  !

I have reviewed and tested this PR as per the instructions, everything works as expected, amazing job! 🌟 🌟 🌟

Question (❓): How do I open a non `wp.com` reader post? This is the only part I didn't test.

I have left minor (🔍) comments, some suggestions (💡)  for you to consider, and a couple of questions (❓). I am going to approve this PR anyway, since none is blocking. I am **NOT** going to merge this PR yet to give you some time to apply any of my suggestions. However, feel free to ignore them and merge the PR yourself.

PS: I loved how you also refactored bits of the existing code and improved both the state of main and test code. Many, many kudos for all this effort! ❤️

ParaskP7 reviewed (commented):
@ashiagr, as per our conversation, I am changing my `Approval` to `Comment` until this the update to the condition to check if previous request is already running is done.

ParaskP7 reviewed (changes requested):
See above comment...

ashiagr committed:
Remove unnecessary braces

ashiagr committed:
Remove braces for when clause "Do Nothing" branches

ashiagr commented:
Thanks for the review @ParaskP7! I've addressed your comments and answered questions.

> How do I open a non wp.com reader post? This is the only part I didn't test.

I tested it by following an external feed (http://timesofindia.indiatimes.com/rssfeedstopstories.cms) and clicking a post from it.  Hope this helps. I should've mentioned it in the description, apologies 🙏.

Ready for another look 🙇‍♀️.

ParaskP7 commented:
👋 @ashiagr !

> I tested it by following an external feed (http://timesofindia.indiatimes.com/rssfeedstopstories.cms) and clicking a post from it. Hope this helps. I should've mentioned it in the description, apologies 🙏.

Many thanks for this tip! 🙏 

> Ready for another look 🙇‍♀️.

All looks good to me. I am going to approve it now. Feel free to merge it when you are ready on your side as well. 🙏

ashiagr committed:
Remove comment

Comments for related posts types exist in handleRelatedPostsResponse and can be removed at this place.
